### PR TITLE
Feature: support automate function input secrets in C# SDK

### PR DESCRIPTION
**Pull request by cdriesler**
## Description & motivation

Automate functions can flag certain inputs as secrets. This will guarantee that their values are redacted in the UI and in any logging during execution. In order for an input to be considered a secret, its JSONSchema property must have `writeOnly` specified as `true`.

In any given language's SDK, we generate Automate JSONSchemas from a class within the source code. In Python, this is currently done with a [pydantic class](https://github.com/specklesystems/speckle_automate_python_example/blob/4b9d6b051d016ca84cb5971d63e4122ababcd1b7/main.py#L25) `SecretStr`, but the built in `DataAnnotations` we use in C# do not provide a way to generate `writeOnly`.

For "extension" cases like this, Newtonsoft provides the ability to implement and attach our own serialization logic, and so we've reached for that to support secrets in the C# SDK.

## Changes:

- Adds `SecretAttribute` annotation to the C# SDK
- Checks for its presence during a custom `JSchemaGenerationProvider` and sets `writeOnly` to true on the generated (in-flight generation) schema

## Validation of changes:

- Used in [this branch](https://github.com/specklesystems/SpeckleAutomateDotnetExample/blob/secret-annotation/SpeckleAutomateDotnetExample/FunctionInputs.cs) of the SpeckleAutomateDotnetExample and verified generated schema includes `writeOnly`
- We do not currently test this logic. Am happy to extract it/add some tests as a prerequisite for this PR but wanted to avoid a change-and-refactor situation.

## References

- [Discussion on forums](https://speckle.community/t/how-do-i-pass-secrets-to-my-automate-function/10356/2?u=iainsproat)
- [Create a custom JSchemaGenerationProvider](https://www.newtonsoft.com/jsonschema/help/html/CreateCustomProvider.htm)

### CODE REVIEW

**Review by gjedlicska — APPROVED** [review]
lgtm

**Comment by AlanRynne** [comment]
Hey @cdriesler! I noticed your PR is failing on the formatting checks, not sure if we've run you through the setup yet but maybe we should.

Meanwhile, you can just:

- `dotnet tools restore`
- `dotnet csharpier ./`

Which should run the right formatter for you ;)

**Comment by cdriesler** [comment]
@AlanRynne ty! that did it. will definitely be finding you again for setup tips

@gjedlicska I had to make some changes for formatting but also for a strange typing on `JSchemaGenerationProvider`. (The `#pragma` bit). Re-requested review.

**Review by gjedlicska — APPROVED** [review]

**Comment by AlanRynne** [comment]
Has the timeline to release this been discussed? Should it go into `main` and be hotfixed?

We're already days away from releasing 2.19 so if there is no 🔥 to put out let's target this into `dev` instead.

**Review by JR-Morgan — APPROVED** [review]

File: Automate/Speckle.Automate.Sdk/Runner.cs [hunk 1]
```diff
@@ -144,3 +147,34 @@ public static async Task<int> Main<TInput>(string[] args, Func<AutomationContext
     return returnCode;
   }
 }
+
+public class SpeckleSecretProvider : JSchemaGenerationProvider
+{
+  // `GetSchema` returning `null` indicates that the given type should not have a customised schema
+  // Nullability of JSchemaTypeGenerationContext appears to be incorrect.
+#pragma warning disable CS8764 // Nullability of return type doesn't match overridden member (possibly because of nullability attributes).
```
Comment (JR-Morgan) [inline Automate/Speckle.Automate.Sdk/Runner.cs:155]: I've opened an issue in the Netwonsoft repo, as the nullability of `JSchemaGenerationProvider.GetSchema` appears incorrect https://github.com/JamesNK/Newtonsoft.Json.Schema/issues/341

**Pull request merged**